### ISSUE TITLE
Add VideoView backed by TextureView

### ISSUE THREAD

**Issue by tzafrir** [twilio/video-quickstart-android#149]
### Description

When I test application on Nexus 6P I420Frame I have to use OES texture to get yuvPlanes using YuvConverter.
But something strange is happen when I try to use this converter for different views (first is local, second is remote).
Example:
```java
public class TextureViewRenderer implements VideoRenderer {
	public void renderFrame(I420Frame frame) {
		// I can only use YuvConverter for localView in this place.
		// If I use it for remoteView there is exception on compile shader: java.lang.RuntimeException: glCreateShader() failed. GLES20 error: 0
		// I need to use YuvConverter only in GL thread for remoteView after that there are no one exceptions anymore.
		// But anyway, I don't see remote stream, all frames are black, as if OES texture contains nothing.
		// Although, I can see local stream from camera that using identical VideoRenderer.
	}
}

// Creating local renderer
TextureViewRenderer localView;

localAudioTrack = LocalAudioTrack.create(this, true);
cameraCapturerCompat = new CameraCapturerCompat(this, 		CameraCapturer.CameraSource.FRONT_CAMERA);
localVideoTrack = LocalVideoTrack.create(this, true, 			cameraCapturerCompat.getVideoCapturer());
localView.setMirror(true);
localVideoTrack.addRenderer(localView);

// Creating remote renderer
TextureViewRenderer remoteView;

private Room.Listener roomListener(final String participantName) {
	return new Room.Listener() {
		@Override
		public void onConnected(Room room) {
			for (Participant participant : room.getParticipants()) {
				if (participant.getVideoTracks().size() > 0) {
					VideoTrack videoTrack = participant.getVideoTracks().get(0);
					remoteView.setMirror(false);
					videoTrack.addRenderer(remoteView);
				}
			}
		}
		...
	}
}
```

And also, if I use another device where using frame.yuvPlanes everything is working excelent.

#### Expected Behavior

See remote video

#### Actual Behavior

Black frame

#### Reproduces how Often

100%

#### Video Android SDK

1.3.0

#### Android API

25

#### Android Device
Nexus 6P

**Comment by aaalaniz** [comment]
Hey @tzafrir 

This sounds like an interesting bug. Can you share how you are using `YuvConverter`? Also is your intent to implement a renderer based on `TextureView` instead of `SurfaceView`? I know `TextureView` is easier to use in a layout, so it makes sense that we add another renderer based on `TextureView`.

**Comment by aaalaniz** [comment]
Also you can reference our [custom renderer example](https://github.com/twilio/video-quickstart-android/blob/master/exampleCustomVideoRenderer/src/main/java/com/twilio/video/examples/customrenderer/SnapshotVideoRenderer.java#L72) for proper use of YuvConverter.

**Comment by tzafrir** [comment]
Yes, I have to use TextureView for renderer where using complex shader to draw final frames.
I can't use a SurfaceView for my renderer because a layout contains complex hierarchy of different elements and TextureView elements among them with overlay views.
A SurfaceView doesn't allow to implement that.

There is example how I use the YuvConverter:

```java
// Example
ByteBuffer outputFrameBuffer = null;
YuvConverter yuvConverter = null;

int outputFrameSize = width * height * 3 / 2;
if (this.outputFrameBuffer == null) {
    this.outputFrameBuffer = ByteBuffer.allocateDirect(outputFrameSize);
}
if (this.yuvConverter == null) {
    this.yuvConverter = new YuvConverter();
}
this.yuvConverter.convert(this.outputFrameBuffer,
        width, height, stride,
        frame.textureId,
        texMatrix);

if (this.outputFrameBuffer != null) {
    ByteBuffer[] yuvPlanes;
    int[] yuvStrides;

    i = this.layoutLock;
    synchronized (this.layoutLock) {
        byte[] data = this.outputFrameBuffer.array();
        int offset = this.outputFrameBuffer.arrayOffset();

        yuvPlanes = new ByteBuffer[]{
                ByteBuffer.allocateDirect(width * height),
                ByteBuffer.allocateDirect(width * height / 4),
                ByteBuffer.allocateDirect(width * height / 4)
        };

        yuvStrides = new int[]{
                width,
                (width + 1) / 2,
                (width + 1) / 2
        };

        yuvPlanes[0].put(data, offset, width * height);

        for (int r = height; r < height * 3 / 2; ++r) {
            yuvPlanes[1].put(data, offset + r * stride, stride / 2);
        }

        for (int r = height; r < height * 3 / 2; ++r) {
            yuvPlanes[2].put(data, offset + r * stride + stride / 2, stride / 2);
        }
    }
}

.....```

**Comment by aaalaniz** [comment]
Hey @tzafrir 

Sorry for the delayed response. Unfortunately, I do not see any issue with the use YuvConverter. Can you provide more details on how the TextureView is implemented? How are you directing frames to the TextureView?

I will mark this issue as an enhancement because it would be valuable to have a VideoView that is based on a TextureView.

Thanks!

**Comment by tzafrir** [comment]
Here's an example of how we implemented TextureView and GlDrawer, hopefully helpful for this issue: https://gist.github.com/tzafrir/772a62c861ac603e8fdbac1738030bde

**Comment by tzafrir** [comment]
Our use case is of showing multiple participants in the room along with other UI elements such as photos and text elements.

If I am understanding correctly, this is unsupported in existing SDK?

Do you have a list of devices that must use surface texture instead of memory buffer? So we can exclude our app from those devices and have a working release, otherwise we cannot use Twilio at all in any of our products.
As of right now we are only aware of this issue on Nexus6P

**Comment by aaalaniz** [comment]
Hey @tzafrir 

Allow me to clarify a few things:

1. The `VideoView` that ships with our SDK does allow for other UI elements along side in the view hierarchy (TextView, ImageView, etc). However, we have not tested the VideoView alongside elements such as a TextureView or other SurfaceViews, so you may be experiencing problems with your particular use case. If the `VideoView` prohibits you from implementing your app's desired view hierarchy then we need to capture this as a separate bug, so I can list the problem as a known issue.

2. As of the `1.3.0` release, all local video is decoded and rendered from a memory buffer and remote video is decoded and rendered from a texture. We noticed some issues with our `CameraCapturer` when trying to encode frames to a texture so we have not enabled this feature for local video. In short, any video rendered from a `LocalVideoTrack` created with `CameraCapturer` uses a memory buffer and the remote video received from a participant's `VideoTrack` uses a texture.

3. I apologize for missing the note on the issue only occurring on the Nexus 6p. I have this device and will investigate using the code samples you shared to see if I can diagnose the issue with black frames you are seeing.

4. We have a task in our backlog for implementing a `TextureView` based renderer. I can't provide timelines but will provide updates on this issue when I have any updates.

5. Can you please share logs when you see black frames? Thanks

Thanks for writing in and sharing your code.

**Comment by tzafrir** [comment]
Log attached:

[nexus6p-log.txt](https://github.com/twilio/video-quickstart-android/files/1298660/nexus6p-log.txt)

**Comment by aaalaniz** [comment]
Hey @tzafrir 

I appreciate your patience. I am investigating this issue the rest of the week. I will provide updates as I make progress.

Thanks!

**Comment by aaalaniz** [comment]
Hey @tzafrir 

I'm not able to use the TextureViewRenderer you shared. All I see is white on the screen. However, I was able to implement a VideoView that extends a TextureView instead of a SurfaceView. I'm still doing more testing, but would this be helpful to your efforts or do you require your own custom TextureView renderer? 

I'm currently thinking we ship another view called `VideoTextureView` that extends `TextureView`. What are your thoughts?

**Comment by aaalaniz** [comment]
Hey @tzafrir 

I hacked together a custom renderer that uses a TextureView that you can try in your application. 

[Rendering with a TextureView](https://gist.github.com/aaalaniz/bfbc4891c98ef3b23558ff2260cbcc8e)

There is some reflection involved, but I tried to mitigate the performance impact as much as possible. You should be able to use `VideoTextureView` similarly to `VideoView`. Let me know if this  renderer works for your application. I will take feedback and work on shipping this renderer with our SDK.

Thanks!

**Comment by aaalaniz** [comment]
Hey @tzafrir 

Does the renderer I sent you work for your application? Looking forward to hearing the results.

Thanks!

**Comment by aaalaniz** [comment]
Hey @tzafrir 

Can you provide an update on your progress?

Thanks!

**Comment by tzafrir** [comment]
We have now successfully completed the project, thanks to your code sample!
Using EglRenderer was useful.

Thanks!

**Comment by aaalaniz** [comment]
Great news @tzafrir !

I will work on adding a TextureView based renderer to our SDK. Please let me know you find any issues with the sample renderer I provided. That renderer will be the baseline for the renderer added to the SDK.

**Comment by smithaaron** [comment]
@aaalaniz Any update on adding a TextureView to the SDK? 

I need to create a round VideoView (it also needs to be placed on top of another full-screen VideoView). My initial research indicates that this can only be done with a TextureView and isn't possible with a SurfaceView. If you have any pointers to how this could be implemented (or if it's possible) with Twilio, that would be appreciated.

**Comment by idelgado** [comment]
@smithaaron we have not formally added this work in our SDK. For now you can use the sample that @aaalaniz linked above: https://gist.github.com/aaalaniz/bfbc4891c98ef3b23558ff2260cbcc8e. TextureView based rendering is in our backlog but we don't have a firm date for completing this work yet.

**Comment by BdeDev** [comment]
Hey @aaalaniz  i want to stream a video with twillio ,but i am unable to do so,with the Help of Your Custom Video Capture Example i came to know about the UI elements can be Streamed ,but the Video is showing Black on Renderer Side.

**Comment by frize** [comment]
I meet the same problem as @tzafrir.
However, my goal is not to draw on a TextureView but I need to get the yuvPlanes buffer for further processing (in particular, I want to get the yuvPlanes buffer to pass it to a C++ buffer). So I choose to follow [this link](https://github.com/twilio/video-quickstart-android/blob/master/exampleCustomVideoRenderer/src/main/java/com/twilio/video/examples/customrenderer/SnapshotVideoRenderer.java#L72) suggested by @aaalaniz

The problem is I all the pixels obtained from YuvConverter.convert are black.
I confirmed using the following code:

```
YuvConverter yuvConverter = new YuvConverter();
yuvConverter.convert(outputFrameBuffer,
                     width,
                     height,
                     width,
                     i420Frame.textureId,
                     texMatrix);

// Now we need to unpack the YUV data into planes
byte[] data = outputFrameBuffer.array();
boolean hasNonZero = false;
for (int i=0; i<data.length; i++){
    if (data[i] > 0){
        Log.d(TAG, "Remote Frame Buffer has non zero pixel at " + i + " = " + (int)data[i]);
        hasNonZero = true;
    }
}
if (hasNonZero == false){
    Log.w(TAG, "Remote Frame Buffer is empty while textureId=" + i420Frame.textureId);
}
else{
    Log.d(TAG, "Remote Frame Buffer has data");
}
```

But the log I received always is:
`Remote Frame Buffer is empty while textureId=2`

Do you have any suggestion for my case @aaalaniz?

**Comment by aaalaniz** [comment]
Hey @frize 

I think you are encountering #122. I believe the frame needs to be extracted from the same EGL context that created it. This might explain why you see black frames. There is no public API for getting the EGL context of the frame today, but I can show you how to get it using reflection. See the gist [Rendering with a TextureView](https://gist.github.com/aaalaniz/bfbc4891c98ef3b23558ff2260cbcc8e)

```
// Get EglBaseProvider
Object eglBaseProvider = EglBaseProviderReflectionUtils.getEglBaseProvider(this);

// Get Root Egl Context
EglBase.Context rootEglBaseContext = (EglBase.Context) getEglBaseContextMethod.invoke(rootEglBase);

// Release EglBaseProvider
EglBaseProviderReflectionUtils.relaseEglBaseProvider(eglBaseProvider, this);
```

**Comment by frize** [comment]
Thank you @aaalaniz, it works perfectly.

For who is concerning, I used two functions to Setup and Cleanup OpenGL ES environment:

```
private void SetupOpenGLESIfNeeded() {
        if (myEgl == null) {
            // Get EglBaseProvider
            eglBaseProvider = EglBaseProviderReflectionUtils.getEglBaseProvider(this);
            // Get Root Egl Context
            EglBase.Context rootEglBaseContext = (EglBase.Context) EglBaseProviderReflectionUtils.getRootEglBaseContext(eglBaseProvider);
            myEgl = EglBase14.create(rootEglBaseContext);
            //TODO: check if surface size is matter
            myEgl.createPbufferSurface(640, 480);
            myEgl.makeCurrent();

            yuvConverter = new YuvConverter();
        }
    }
```

```
private void CleanUpOpenGLESIfNeeded() {
        if (myEgl != null) {
            yuvConverter.release();

            myEgl.releaseSurface();
            myEgl.release();
            // Release EglBaseProvider
            EglBaseProviderReflectionUtils.relaseEglBaseProvider(eglBaseProvider, this);
        }
    }
```

Also remember that renderFrame could be called from a different thread from the thread that you created OpenGL ES, so make sure that you run the yuvConverter.convert in the correct thread.

**Comment by aaalaniz** [comment]
Awesome @frize !

We are getting more inbound about manipulating rendered frames. Given that EglBaseProvider is essential to this type of work we will considering making it a public class so that developers can get the egl context without having to use reflection.

**Comment by aaalaniz** [comment]
@zackm0571 released `VideoTextureView` in `2.2.0.` 🎉
